src/charm.py

```diff
@@ -0,0 +1,226 @@
+from loki import (
+    CONTAINER_NAME,
+    LOKI_CERT_FILE,
+    Loki,
+)
+from loki_cluster import (
+    ConfigReceivedEvent,
+    LokiClusterRequirer,
+    LokiRole,
+)
+from ops.charm import CharmBase, CollectStatusEvent
+from ops.framework import BoundEvent, Object
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+
+# Log messages can be retrieved using juju debug-log
+logger = logging.getLogger(__name__)
+
+
+class LokiWorkerK8SOperatorCharm(CharmBase):
+    """A Juju Charmed Operator for Loki."""
+
+    _loki = None
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._container = self.unit.get_container(CONTAINER_NAME)
+        self._loki = Loki(self._container, self._loki_roles, self.model)
+        self._topology = JujuTopology.from_charm(self)
+        self._loki_cluster = LokiClusterRequirer(self)
+        self._log_forwarder = ManualLogForwarder(
+            charm=self,
+            loki_endpoints=self._loki_cluster.get_loki_endpoints(),
+            refresh_events=[
+                self.on["loki-cluster"].relation_joined,
+                self.on["loki-cluster"].relation_changed,
+            ],
+        )
+        self.unit.set_ports(self._loki.port)
+
+        # === EVENT HANDLER REGISTRATION === #
+        self.framework.observe(self.on.loki_pebble_ready, self._on_pebble_ready)  # pyright: ignore
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_status)
+        self.framework.observe(
+            self._loki_cluster.on.config_received, self._on_loki_config_received
+        )
+        self.framework.observe(self._loki_cluster.on.created, self._on_loki_cluster_created)
+        self.framework.observe(
+            self.on.loki_cluster_relation_departed, self._log_forwarder.disable_logging
+        )
+        self.framework.observe(
+            self.on.loki_cluster_relation_broken, self._log_forwarder.disable_logging
+        )
+
+    # === EVENT HANDLERS === #
+    def _on_pebble_ready(self, _):
+        self.unit.set_workload_version(
+            self._loki.version or ""  # pyright: ignore[reportOptionalMemberAccess]
+        )
+        self._update_config()
+
+    def _on_config_changed(self, _):
+        # if the user has changed the roles, we might need to let the coordinator know
+        self._update_loki_cluster()
+
+        # if we have a config, we can start loki
+        if self._loki_cluster.get_loki_config():
+            # determine if a workload restart is necessary
+            self._update_config()
+
+    def _on_upgrade_charm(self, _):
+        self._update_loki_cluster()
+
+    def _on_collect_status(self, e: CollectStatusEvent):
+        if not self._container.can_connect():
+            e.add_status(WaitingStatus("Waiting for `loki` container"))
```

Comment:
There is a potential race here. To be consistent throughout, we need to check `can_connect` once in `_configure`, otherwise it could be False there but True here. [Example](https://github.com/canonical/prometheus-k8s-operator/blob/20b5d45e7c2c9b1fc5628b2878bdf492526f206a/src/charm.py#L539).